### ISSUE TITLE
dhcp_relay failed on current master

### ISSUE THREAD

**Issue by justindthomas** [sonic-net/sonic-buildimage#17379]
#### Description

Currently I'm running SONiC with a build date of 11/20. I upgraded to the current master build (as of 12/01) this afternoon. I immediately noticed that my WiFi devices would not come online. After some investigating, I found that the `dhcp_relay` container failed to start with these errors:

```
root@sonic:~# docker ps -a
CONTAINER ID   IMAGE                                COMMAND                  CREATED          STATUS                      PORTS     NAMES
1f791971baf6   docker-snmp:latest                   "/usr/local/bin/supe…"   24 minutes ago   Up 24 minutes                         snmp
da66859731ba   docker-platform-monitor:latest       "/usr/bin/docker_ini…"   24 minutes ago   Up 24 minutes                         pmon
ac4329f73dea   docker-sonic-mgmt-framework:latest   "/usr/local/bin/supe…"   24 minutes ago   Up 24 minutes                         mgmt-framework
2bffd06fe095   docker-lldp:latest                   "/usr/bin/docker-lld…"   25 minutes ago   Up 25 minutes                         lldp
53a3e0a4dfac   docker-sonic-gnmi:latest             "/usr/local/bin/supe…"   25 minutes ago   Up 25 minutes                         gnmi
cb1e6d70e998   443a3ed104c6                         "/usr/bin/docker_ini…"   26 minutes ago   Exited (2) 23 minutes ago             dhcp_relay
6e9c6a000acc   docker-router-advertiser:latest      "/usr/bin/docker-ini…"   26 minutes ago   Up 26 minutes                         radv
633e1fca33fc   docker-fpm-frr:latest                "/usr/bin/docker_ini…"   26 minutes ago   Up 26 minutes                         bgp
5468ad1aef13   docker-syncd-brcm:latest             "/usr/local/bin/supe…"   26 minutes ago   Up 26 minutes                         syncd
9ebe741e10f3   docker-teamd:latest                  "/usr/local/bin/supe…"   26 minutes ago   Up 26 minutes                         teamd
580685dc0005   docker-orchagent:latest              "/usr/bin/docker-ini…"   27 minutes ago   Up 26 minutes                         swss
a6434d616dfd   docker-eventd:latest                 "/usr/local/bin/supe…"   27 minutes ago   Up 26 minutes                         eventd
cabb3c5cffec   docker-database:latest               "/usr/local/bin/dock…"   27 minutes ago   Up 27 minutes                         database
root@sonic:~# docker logs dhcp_relay
/usr/local/lib/python3.9/dist-packages/supervisor/options.py:473: UserWarning: Supervisord is running as root and it is searching for its configuration file in default locations (including its current working directory); you probably want to specify a "-c" argument specifying an absolute path to a configuration file for improved security.
  self.warnings.warn(
Error: [program:dhcprelayd] names unknown program or fcgi-program
For help, use /usr/local/bin/supervisord -h
```

I rolled back to the previous version of SONiC I was running. I can switch back on Monday if specific details are required for diagnosing.

I also noticed that the new version is running Debian 12 (awesome!) which likely carries a significant amount of changes from the last-known-good Debian 11 version.

**Comment by justindthomas** [comment]
I tried tweaking a couple of things in the container and building from source.B ut now that I'm testing with the container live, I noticed that the templates aren't reflecting the current data in the repository. It looks like maybe the build process just isn't including the updated `dhcp_relay` container at all:

```
SONiC Software Version: SONiC.master.0-dirty-20231202.084620
SONiC OS Version: 12
Distribution: Debian 12.2
Kernel: 6.1.0-11-2-amd64
Build commit: 8db3a99d1
Build date: Sat Dec  2 10:14:28 UTC 2023
Built by: justin@sonic-build

Platform: x86_64-dellemc_n3248te_c3338-r0
HwSKU: DellEMC-N3248TE
ASIC: broadcom
ASIC Count: 1
Serial Number: 4GNXV43
Model Number: 0WNWT9
Hardware Revision:
Uptime: 17:15:20 up 36 min,  2 users,  load average: 2.42, 2.60, 3.53
Date: Sat 02 Dec 2023 17:15:20

Docker images:
REPOSITORY                    TAG                              IMAGE ID       SIZE
docker-dhcp-relay             latest                           8bf9f58ca896   312MB
docker-macsec                 latest                           449e90529799   330MB
docker-eventd                 latest                           6d19a3b1b207   302MB
docker-eventd                 master.0-dirty-20231202.084620   6d19a3b1b207   302MB
docker-snmp                   latest                           0ca6f8b18d8f   341MB
docker-snmp                   master.0-dirty-20231202.084620   0ca6f8b18d8f   341MB
docker-sflow                  latest                           c538ddffe175   329MB
docker-sflow                  master.0-dirty-20231202.084620   c538ddffe175   329MB
docker-platform-monitor       latest                           53eded32ad5b   422MB
docker-platform-monitor       master.0-dirty-20231202.084620   53eded32ad5b   422MB
docker-fpm-frr                latest                           b4ea3164a183   359MB
docker-fpm-frr                master.0-dirty-20231202.084620   b4ea3164a183   359MB
docker-teamd                  latest                           57ae3cf75341   328MB
docker-teamd                  master.0-dirty-20231202.084620   57ae3cf75341   328MB
docker-sonic-gnmi             latest                           2a56a996ef5c   390MB
docker-sonic-gnmi             master.0-dirty-20231202.084620   2a56a996ef5c   390MB
docker-orchagent              latest                           c197c519a344   340MB
docker-orchagent              master.0-dirty-20231202.084620   c197c519a344   340MB
docker-syncd-brcm             latest                           03392ca594de   677MB
docker-syncd-brcm             master.0-dirty-20231202.084620   03392ca594de   677MB
docker-nat                    latest                           df6f3a50ee56   331MB
docker-nat                    master.0-dirty-20231202.084620   df6f3a50ee56   331MB
docker-lldp                   latest                           7dd48f614023   345MB
docker-lldp                   master.0-dirty-20231202.084620   7dd48f614023   345MB
docker-mux                    latest                           85f13fc2aff3   351MB
docker-mux                    master.0-dirty-20231202.084620   85f13fc2aff3   351MB
docker-gbsyncd-broncos        latest                           977a0c59596f   353MB
docker-gbsyncd-broncos        master.0-dirty-20231202.084620   977a0c59596f   353MB
docker-gbsyncd-credo          latest                           62bfa5492881   325MB
docker-gbsyncd-credo          master.0-dirty-20231202.084620   62bfa5492881   325MB
docker-router-advertiser      latest                           d77b5cfa1192   302MB
docker-router-advertiser      master.0-dirty-20231202.084620   d77b5cfa1192   302MB
docker-database               latest                           6b9ae4af4faa   302MB
docker-database               master.0-dirty-20231202.084620   6b9ae4af4faa   302MB
docker-sonic-mgmt-framework   latest                           c676394b035d   418MB
docker-sonic-mgmt-framework   master.0-dirty-20231202.084620   c676394b035d   418MB
```

**Comment by justindthomas** [comment]
It looks to me like `docker-dhcp-relay` may not be getting built. In my working system from 11/20, `docker-dhcp-relay` is the only container without a `master.uuid` tag. I attempted to tweak the templates for dhcp-relay and build the whole system myself and see the same behavior (i.e., all of the other containers show the `master.uuid-dirty` tags, but `docker-dhcp-relay` does not (and my changes aren't reflected in the built system).

From the 11/20 system:

```
jdt@sonic:~$ sudo show ver

SONiC Software Version: SONiC.master.415079-b93852d53
SONiC OS Version: 11
Distribution: Debian 11.7
Kernel: 5.10.0-23-2-amd64
Build commit: b93852d53
Build date: Mon Nov 20 13:02:14 UTC 2023
Built by: AzDevOps@vmss-soni002HLC

Platform: x86_64-dellemc_n3248te_c3338-r0
HwSKU: DellEMC-N3248TE
ASIC: broadcom
ASIC Count: 1
Serial Number: 4GNXV43
Model Number: 0WNWT9
Hardware Revision:
Uptime: 18:24:32 up  1:07,  1 user,  load average: 2.74, 2.36, 2.32
Date: Sat 02 Dec 2023 18:24:32

Docker images:
REPOSITORY                    TAG                       IMAGE ID       SIZE
docker-gbsyncd-broncos        latest                    6fd81670cbd4   351MB
docker-gbsyncd-broncos        master.415079-b93852d53   6fd81670cbd4   351MB
docker-gbsyncd-credo          latest                    1741e599df6a   323MB
docker-gbsyncd-credo          master.415079-b93852d53   1741e599df6a   323MB
docker-syncd-brcm             latest                    45ff8546690e   675MB
docker-syncd-brcm             master.415079-b93852d53   45ff8546690e   675MB
docker-orchagent              latest                    4ca77939c4cb   338MB
docker-orchagent              master.415079-b93852d53   4ca77939c4cb   338MB
docker-teamd                  latest                    9dc4ba683268   326MB
docker-teamd                  master.415079-b93852d53   9dc4ba683268   326MB
docker-fpm-frr                latest                    ecda7924d0e7   357MB
docker-fpm-frr                master.415079-b93852d53   ecda7924d0e7   357MB
docker-dhcp-relay             latest                    a7f233e8a8e1   309MB
docker-sflow                  latest                    b5d9b6ace420   327MB
docker-sflow                  master.415079-b93852d53   b5d9b6ace420   327MB
docker-nat                    latest                    2446df575893   329MB
docker-nat                    master.415079-b93852d53   2446df575893   329MB
docker-macsec                 latest                    4b9fc9eb9a3b   328MB
docker-eventd                 latest                    0a7c16f959db   300MB
docker-eventd                 master.415079-b93852d53   0a7c16f959db   300MB
docker-platform-monitor       latest                    f2fe4c97b942   421MB
docker-platform-monitor       master.415079-b93852d53   f2fe4c97b942   421MB
docker-snmp                   latest                    6b6bb79a2a91   340MB
docker-snmp                   master.415079-b93852d53   6b6bb79a2a91   340MB
docker-router-advertiser      latest                    f006c8455a2c   301MB
docker-router-advertiser      master.415079-b93852d53   f006c8455a2c   301MB
docker-lldp                   latest                    b1c5e51a729f   343MB
docker-lldp                   master.415079-b93852d53   b1c5e51a729f   343MB
docker-mux                    latest                    a54a207ff68c   349MB
docker-mux                    master.415079-b93852d53   a54a207ff68c   349MB
docker-database               latest                    e58cea41882a   301MB
docker-database               master.415079-b93852d53   e58cea41882a   301MB
docker-sonic-gnmi             latest                    8065e3f5fa45   388MB
docker-sonic-gnmi             master.415079-b93852d53   8065e3f5fa45   388MB
docker-sonic-mgmt-framework   latest                    c6848441dcda   416MB
docker-sonic-mgmt-framework   master.415079-b93852d53   c6848441dcda   416MB
```

...and from my local build system using the current master branch:

```
jdt@sonic:~$ sudo show ver

SONiC Software Version: SONiC.master.0-dirty-20231202.084620
SONiC OS Version: 12
Distribution: Debian 12.2
Kernel: 6.1.0-11-2-amd64
Build commit: 8db3a99d1
Build date: Sat Dec  2 10:14:28 UTC 2023
Built by: justin@sonic-build

Platform: x86_64-dellemc_n3248te_c3338-r0
HwSKU: DellEMC-N3248TE
ASIC: broadcom
ASIC Count: 1
Serial Number: 4GNXV43
Model Number: 0WNWT9
Hardware Revision:
Uptime: 17:15:20 up 36 min,  2 users,  load average: 2.42, 2.60, 3.53
Date: Sat 02 Dec 2023 17:15:20

Docker images:
REPOSITORY                    TAG                              IMAGE ID       SIZE
docker-dhcp-relay             latest                           8bf9f58ca896   312MB
docker-macsec                 latest                           449e90529799   330MB
docker-eventd                 latest                           6d19a3b1b207   302MB
docker-eventd                 master.0-dirty-20231202.084620   6d19a3b1b207   302MB
docker-snmp                   latest                           0ca6f8b18d8f   341MB
docker-snmp                   master.0-dirty-20231202.084620   0ca6f8b18d8f   341MB
docker-sflow                  latest                           c538ddffe175   329MB
docker-sflow                  master.0-dirty-20231202.084620   c538ddffe175   329MB
docker-platform-monitor       latest                           53eded32ad5b   422MB
docker-platform-monitor       master.0-dirty-20231202.084620   53eded32ad5b   422MB
docker-fpm-frr                latest                           b4ea3164a183   359MB
docker-fpm-frr                master.0-dirty-20231202.084620   b4ea3164a183   359MB
docker-teamd                  latest                           57ae3cf75341   328MB
docker-teamd                  master.0-dirty-20231202.084620   57ae3cf75341   328MB
docker-sonic-gnmi             latest                           2a56a996ef5c   390MB
docker-sonic-gnmi             master.0-dirty-20231202.084620   2a56a996ef5c   390MB
docker-orchagent              latest                           c197c519a344   340MB
docker-orchagent              master.0-dirty-20231202.084620   c197c519a344   340MB
docker-syncd-brcm             latest                           03392ca594de   677MB
docker-syncd-brcm             master.0-dirty-20231202.084620   03392ca594de   677MB
docker-nat                    latest                           df6f3a50ee56   331MB
docker-nat                    master.0-dirty-20231202.084620   df6f3a50ee56   331MB
docker-lldp                   latest                           7dd48f614023   345MB
docker-lldp                   master.0-dirty-20231202.084620   7dd48f614023   345MB
docker-mux                    latest                           85f13fc2aff3   351MB
docker-mux                    master.0-dirty-20231202.084620   85f13fc2aff3   351MB
docker-gbsyncd-broncos        latest                           977a0c59596f   353MB
docker-gbsyncd-broncos        master.0-dirty-20231202.084620   977a0c59596f   353MB
docker-gbsyncd-credo          latest                           62bfa5492881   325MB
docker-gbsyncd-credo          master.0-dirty-20231202.084620   62bfa5492881   325MB
docker-router-advertiser      latest                           d77b5cfa1192   302MB
docker-router-advertiser      master.0-dirty-20231202.084620   d77b5cfa1192   302MB
docker-database               latest                           6b9ae4af4faa   302MB
docker-database               master.0-dirty-20231202.084620   6b9ae4af4faa   302MB
docker-sonic-mgmt-framework   latest                           c676394b035d   418MB
docker-sonic-mgmt-framework   master.0-dirty-20231202.084620   c676394b035d   418MB
```

I'm reading up on README.buildsystem.md but so far haven't found anything that dictates a list of what containers are built. I'll keep digging, but if someone can offer a pointer to the right file, I'd appreciate it.

**Comment by justindthomas** [comment]
My simplified and updated version of `dhcp-relay.programs.j2` looks like this:

```jinja2
[group:dhcp-relay]
programs=dhcprelayd
{% set relay_for_ipv6 = { 'flag': False } %}
{% for vlan_name in VLAN_INTERFACE %}
{% if DHCP_RELAY and vlan_name in DHCP_RELAY and DHCP_RELAY[vlan_name]['dhcpv6_servers']|length > 0 %}
{% set _dummy = relay_for_ipv6.update({ 'flag': True }) %}
{% endif %}
{% endfor %}
{# Append DHCPv6 agents #}
{% if relay_for_ipv6.flag -%}
,dhcp6relay
{% endif %}
```

I'm assuming that `dhcprelayd` is necessary for both IPv4 and IPv6-only deployments. I'll need to adjust the logic if only `dhcp6relay` is needed for IPv6-only scenarios.

**Comment by justindthomas** [comment]
Resolved in #17506
